### PR TITLE
Adds examples/pigweed-app

### LINKED ISSUE #2780: Add pigweed RPC example app
#### Problem
Example app integrated with third_party/pigweed
running on nRF52840, in preparation to on-device
testing.

 #### Proposed Solution
Create example app, integrate with third_party/pigweed

### PR DESCRIPTION
**Pull request by szatmz**
#### Problem
Example app integrated with third_party/pigweed
running on nRF52840.
Compilation needs the NRF SDK and tools installed
and the respective environment variables set.

 #### Summary of Changes
Adds examples/pigweed_app and its dependencies,
examples/platform/nrf528xx/pw_sys_io, a pw_sys_io
implementation over APP_UARTE in the NRF SDK, and
examples/platform/nrf528xx/sdk, a common nrf5_sdk
target.

Once the image is built and flashed, the interactive
Pigweed RPC IPython environment can be started like this:

    python -m pw_hdlc_lite.rpc_console --device /dev/ttyACM0 -b 115200 -p ../../../third_party/pigweed/repo/pw_rpc/pw_rpc_protos/echo.proto -o /tmp/pw_rpc.out

where -p is the location of the Echo RPCs proto and -o is
the file name where pw_log messages are sent.

The example pigweed_app has one RPC, Echo, it can be invoked like this:

    rpcs.pw.rpc.EchoService.Echo(msg="hi")

fixes #2780

### CODE REVIEW

**Review by andy31415 — COMMENTED** [review]
A README.md of what 'pigweed app' means would be nice.

I believe we have extra understanding of what pigweed is, however people checking out the repo will have a harder time understanding it (as opposed to say 'lighting-demo' or 'lock-demo' which seem a bit more self-explanatory).

We should also place the instructions of 'how to flash/run/use' in there rather than just PR description.

**Comment by szatmz** [comment]
> A README.md of what 'pigweed app' means would be nice.
> 
> I believe we have extra understanding of what pigweed is, however people checking out the repo will have a harder time understanding it (as opposed to say 'lighting-demo' or 'lock-demo' which seem a bit more self-explanatory).
> 
> We should also place the instructions of 'how to flash/run/use' in there rather than just PR description.

Done

**Comment by mspang** [comment]
Is pigweed app the right name? Maybe it should be named after whatever it's going to do? test rpc server?

**Review by rwalker-apple — COMMENTED** [review]
will you add this to CI?  what does this "app" do?

**Comment by szatmz** [comment]
> Is pigweed app the right name? Maybe it should be named after whatever it's going to do? test rpc server?

Well, I was thinking this app could be the playground for trying all sorts of pigweed functionality in CHIP. Kind of like a guinea pig of Pigweed integration. Guineapigweed :)

**Comment by szatmz** [comment]
> will you add this to CI? what does this "app" do?

Yes, in a new PR I want to add this to .github/workflows/examples.yaml

This app currently serves a simple Echo Pigweed RPC request. Next step is to plumb in actual CHIP functionality that can be exercised through other RPCs, for purposes of on-device testing.

**Comment by rwalker-apple** [comment]
> > will you add this to CI? what does this "app" do?
> 
> Yes, in a new PR I want to add this to .github/workflows/examples.yaml
> 
> This app currently serves a simple Echo Pigweed RPC request. Next step is to plumb in actual CHIP functionality that can be exercised through other RPCs, for purposes of on-device testing.

would it be more appropriate to put pigweed subdirectories in existing examples?

**Comment by szatmz** [comment]
> > > will you add this to CI? what does this "app" do?
> > 
> > 
> > Yes, in a new PR I want to add this to .github/workflows/examples.yaml
> > This app currently serves a simple Echo Pigweed RPC request. Next step is to plumb in actual CHIP functionality that can be exercised through other RPCs, for purposes of on-device testing.
> 
> would it be more appropriate to put pigweed subdirectories in existing examples?

Not sure. Existing example directories are subdivided by manufacturer or architecture. Pigweed is neither. It is mostly a hardware abstraction framework. Currently the aim is to use its RPC feature, but in the future other features may prove useful to integrate into CHIP.

File: examples/platform/nrf528xx/pw_sys_io/BUILD [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2020 The Pigweed Authors
```
Comment (mspang) [inline examples/platform/nrf528xx/pw_sys_io/BUILD:1]: CHIP authors?

File: examples/platform/nrf528xx/pw_sys_io/BUILD.gn [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# Copyright 2019 The Pigweed Authors
```
Comment (mspang) [inline examples/platform/nrf528xx/pw_sys_io/BUILD.gn:1]: CHIP authors?

File: gn/build/config/compiler/BUILD.gn [hunk 3]
```diff
@@ -136,15 +136,19 @@ config("optimize_default") {
...
-    "-Wno-unknown-warning-option",
     "-Wno-missing-field-initializers",
-    "-Wno-unused-but-set-variable",
     "-Wno-unused-variable",
   ]
-  cflags_cc = [
-    "-Wno-non-virtual-dtor",
-    "-Wno-deprecated-copy",
-  ]
+  if (!_is_embedded_arm) {
```
Comment (mspang) [inline gn/build/config/compiler/BUILD.gn:142]: This change breaks the build.

**Review by mspang — CHANGES_REQUESTED** [review]
The copyrights look out of spec.

File: gn/build/config/compiler/BUILD.gn (hunk 3 above)
Comment (szatmz) [reply gn/build/config/compiler/BUILD.gn:142]: Sorry, which build does it break? I had to add it to _un_break android build that's one of the checks.

Comment (mspang) [reply gn/build/config/compiler/BUILD.gn:142]: I think you can drop all of these warnings changes completely.

File: examples/platform/nrf528xx/pw_sys_io/BUILD (hunk 1 above)
Comment (szatmz) [reply examples/platform/nrf528xx/pw_sys_io/BUILD]: done

File: examples/platform/nrf528xx/pw_sys_io/BUILD.gn (hunk 2 above)
Comment (szatmz) [reply examples/platform/nrf528xx/pw_sys_io/BUILD.gn]: done

File: gn/build/config/compiler/BUILD.gn (hunk 3 above)
Comment (szatmz) [reply gn/build/config/compiler/BUILD.gn]: done

File: examples/platform/nrf528xx/pw_sys_io/include/pw_sys_io_baremetal_nrf528xx/init.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
+// Copyright 2020 The Pigweed Authors
```
Comment (mspang) [inline examples/platform/nrf528xx/pw_sys_io/include/pw_sys_io_baremetal_nrf528xx/init.h:1]: CHIP authors?

File: examples/pigweed-app/nrf5/main/main.cpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
+// Copyright 2019 The Pigweed Authors
```
Comment (mspang) [inline examples/pigweed-app/nrf5/main/main.cpp:1]: CHIP authors?

**Review by mspang — APPROVED** [review]

File: examples/platform/nrf528xx/pw_sys_io/include/pw_sys_io_baremetal_nrf528xx/init.h (hunk 4 above)
Comment (szatmz) [reply examples/platform/nrf528xx/pw_sys_io/include/pw_sys_io_baremetal_nrf528xx/init.h]: done

File: examples/pigweed-app/nrf5/main/main.cpp (hunk 5 above)
Comment (szatmz) [reply examples/pigweed-app/nrf5/main/main.cpp]: done

**Review by andy31415 — APPROVED** [review]

**Review by BroderickCarlin — APPROVED** [review]

**Comment by mspang** [comment]
@project-chip/reviewers-apple @project-chip/reviewers-comcast @project-chip/reviewers-amazon

**Review by rwalker-apple — APPROVED** [review]

**Pull request merged**